### PR TITLE
raftstore: move scan delete to raft log gc worker (#11853)

### LINKED ISSUE #10210: Cleaning raft log costs too much time during destroying a peer
### What version of TiKV are you using?

v4.0.6

### What did happened?
Sometimes during destroying a peer, cleaning raft log may cost too much time and cause much high jitter of latency and leaders to drop. :
```
[2021/05/18 10:58:31.160 +08:00] [INFO] [peer_storage.rs:1416] ["finish clear peer meta"] [takes=30.186824947s] [raft_logs=668] [raft_key=1] [apply_key=1] [meta_key=1] [region_id=43002]
```

The root cause could be that [the seek operation](https://github.com/tikv/tikv/blob/363e92dfa900c2c412968ca27a5469c1c6d5c4b3/components/engine_rocks/src/raft_engine.rs#L145) for finding the min log index  may be very slow:
image

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #11853 to release-5.1
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tikv repo:
git pr https://github.com/tikv/tikv/pull/11878
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tikv.git pr/11878:release-5.1-6170255604af
```

---

### What is changed and how it works?

Close #10210.

What's Changed:

When clearing raft metas, raftstore will scan raft logs and delete them
one by one. Seeking can be slow if there are a lot of tombstone keys.
This PR moves the operation to raft log gc worker to reduce the impact.

It offloads the operation by two strategies:
- For destroy, always flush raft gc worker first, so current `last_compacted_idx` is always reliable for cleaning up logs;
- For applying snapshot, use current first index directly, and schedule a raft log gc task after the snapshot is applied. This is a best effort, if TiKV is shutdown before the task is handled, there may still be stale logs. They will be deleted when next compact log task is executed or when the peer is destroyed. This is the same as potential leaking of regular raft gc task.

The final solution should be also moving remaining IO operations to
async write IO threads.

### Check List 

Tests 

- Unit test
- Integration test

### Release note

```release-note
fix potential high latency caused by destroying a peer
```

### CODE REVIEW

**Comment by sticnarf** [comment]
PTAL @5kbpers @NingLin-P

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -131,6 +143,12 @@ where
     max_inflight_msgs: usize,
 
     trace: PeerMemoryTrace,
+    /// Destroy is delayed because of some unpersisted readies in Peer.
+    /// Should call `destroy_peer` again after persisting all readies.
```
Comment (NingLin-P) [inline components/raftstore/src/store/fsm/peer.rs:147]: Plz update the comment

**Review by NingLin-P — APPROVED** [review]

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**